File: shared/golang/api/product.go
package api

import (
	"time"
)

type Product struct {
	ID         int              `json:"id"`
	Context    Context          `json:"context"`
	Attributes ObjectAttributes `json:"attributes"`
	Variants   Variants         `json:"variants"`
	SKUs       []SKU            `json:"skus"`
	Albums     []Album          `json:"albums"`
}

func (p Product) Title() (string, error) {
	value, err := p.Attributes.LookupValue("title")
	return value.(string), err
}

func (p Product) Description() string {
	value, _ := p.Attributes.LookupValue("description")
	return value.(string)
}

func (p Product) FirstImage() string {
	if len(p.Albums) == 0 {
		return ""
	}

	if len(p.Albums[0].Images) == 0 {
		return ""
	}

	return p.Albums[0].Images[0].Source
}

func (p Product) IsActive() bool {
	activeFrom, err := p.Attributes.LookupDateTime("activeFrom")
	if err != nil {
		return false
	}
	activeTo, err := p.Attributes.LookupDateTime("activeTo")
	if err != nil {
		return false
	}
	if activeFrom == nil {
		return false
	}
	return activeFrom.Before(time.Now()) && (activeTo == nil || activeTo.After(time.Now()))
}

func (p Product) Tags() interface{} {
	value, _ := p.Attributes.LookupValue("tags")
	return value
}

File: phoenix-scala/seeder/src/main/scala/gatling/seeds/requests/Product.scala
package gatling.seeds.requests

import java.time.Instant

import gatling.seeds.requests.Auth._
import io.gatling.core.Predef._
import io.gatling.http.Predef._
import models.product._
import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization.{write ⇒ json}
import org.json4s.JsonDSL._
import payloads.ImagePayloads._
import payloads.ProductPayloads._
import payloads.ProductVariantPayloads._
import utils.aliases._
import utils.Money.Currency
import utils.seeds.Factories

case class StringField(value: String, stringType: String = "string") {
  val illuminated: Json = ("t" → stringType) ~ ("v" → value)
}

case class PriceField(currency: Currency, value: Int) {
  val illuminated: Json =
    ("t" → "price") ~ (("currency" → currency.getCode) ~ ("value" → value))
}

object Product {
  private def buildCreatePayload(sp: SimpleProductData): CreateProductPayload = {
    val activeFrom = if (sp.active) s""""${Instant.now}"""" else "null";
    val ts: String = compact(render(JArray(sp.tags.map(t ⇒ JString(t)).toList)))

    val code        = StringField(sp.code).illuminated
    val title       = StringField(sp.title).illuminated
    val description = StringField(sp.description, "richText").illuminated
    val price       = PriceField(sp.currency, sp.price).illuminated
    val activeDate  = StringField(activeFrom, "date-time").illuminated
    val tags        = ("t" → "tags") ~ ("v" → ts)

    val productAttrs =
      Map("title" → title, "description" → description, "activeFrom" → activeDate, "tags" → tags)
    val variantAttrs = Map("code" → code,
                           "title"       → title,
                           "description" → description,
                           "activeFrom"  → activeDate,
                           "salePrice"   → price,
                           "retailPrice" → price)

    val album =
      AlbumPayload(name = Some("default"), images = Some(Seq(ImagePayload(src = sp.image))))

    val variant = ProductVariantPayload(attributes = variantAttrs, albums = Some(Seq(album)))
    CreateProductPayload(attributes = productAttrs,
                         variants = Seq(variant),
                         options = None,
                         albums = Some(Seq(album)))
  }

  private def createNewProduct =
    http("Create product")
      .post("/v1/products/default")
      .requireAdminAuth
      .body(StringBody(session ⇒ session.get("productPayload").as[String]))

  val createProducts = foreach(Factories.products, "simpleProduct") {
    exec(session ⇒ {
      val simple  = session("simpleProduct").as[SimpleProductData]
      val payload = buildCreatePayload(simple)
      session.set("productPayload", payload)
    }).exec(createNewProduct)
  }

}
